packages/ra-core/src/controller/ListController.tsx

```diff
@@ -194,25 +194,34 @@ export class UnconnectedListController extends Component<Props> {
         this.setFilters.cancel();
     }
 
-    componentWillReceiveProps(nextProps: Props) {
+    componentDidUpdate(prevProps: Props) {
+        const {
+            resource,
+            query,
+            params,
+            filter,
+            sort,
+            perPage,
+            version,
+        } = this.props;
+
         if (
-            nextProps.resource !== this.props.resource ||
-            nextProps.query.sort !== this.props.query.sort ||
-            nextProps.query.order !== this.props.query.order ||
-            nextProps.query.page !== this.props.query.page ||
-            nextProps.query.perPage !== this.props.query.perPage ||
-            !isEqual(nextProps.query.filter, this.props.query.filter) ||
-            !isEqual(nextProps.filter, this.props.filter) ||
-            !isEqual(nextProps.sort, this.props.sort) ||
-            !isEqual(nextProps.perPage, this.props.perPage)
+            prevProps.resource !== resource ||
+            prevProps.query.sort !== query.sort ||
+            prevProps.query.order !== query.order ||
+            prevProps.query.page !== query.page ||
+            prevProps.query.perPage !== query.perPage ||
+            !isEqual(prevProps.query.filter, query.filter) ||
+            !isEqual(prevProps.query.filter, query.filter) ||
+            !isEqual(prevProps.params, params) ||
+            !isEqual(prevProps.filter, filter) ||
+            !isEqual(prevProps.sort, sort) ||
+            !isEqual(prevProps.perPage, perPage)
         ) {
-            this.updateData(
-                Object.keys(nextProps.query).length > 0
-                    ? nextProps.query
-                    : nextProps.params
-            );
+            this.updateData(Object.keys(query).length > 0 && query);
```

Comment:
no, you've changed the logic here. Please revert.

Reply:
@fzaninotto this was due to the second bug I mentioned. If you apply the dynamic filter with no query params for example `/#/users` the params with defaults of `null` are called on the update data queries. This throws the `"Cannot read property 'toLowerCase' of null"` error.

Since the params data is merged with the defaults in the `this.getQuery()` I thought that just not passing the params in this case would still work and properly merge the defaults, but perhaps I misunderstood the logic here?

Reply:
with the previous logic, if there is no query string in the URL, we use the params from the Redux store, which are always defined. 

I don't know about the `Cannot read property 'toLowerCase' of null"` error, I suggest we deal with it in another PR, and you provide a failing test case to demonstrate it.